### DISCUSSION TITLE
Does psycopg2 allow passing an existing SSL connection to psycopg2.connect()?

### DISCUSSION THREAD

**Discussion by jackwotherspoon** [psycopg/psycopg2#1420] [Q&A]
The [Cloud SQL Python Connector](https://github.com/GoogleCloudPlatform/cloud-sql-python-connector) would like to support database connections to Cloud SQL using psycopg2. In order to do so we require the ability to pass in an existing connection or the ability to configure the connection level SSL.

Is this something that is currently possible with psycopg2?

For reference, we currently support pg8000 through the use of their `ssl_context` argument which allows us to pass in our pre-configured ssl.SSLContext object. [pg8000 ssl_context](https://github.com/tlocke/pg8000#pg8000-native-connectionuser-hostlocalhost-databasenone-port5432-passwordnone-source_addressnone-unix_socknone-ssl_contextnone-timeoutnone-tcp_keepalivetrue-application_namenone-replicationnone)

**Comment by dvarrazzo** [comment]
Hello Jack,

psycopg (2 and 3) use [the libpq](https://www.postgresql.org/docs/current/libpq.html), the PostgreSQL client library, to connect to the database. I don't believe that the libpq offers such functionality.

Maybe it's something that can be added, but it should be in the libpq; once it's available we would be happy to support it in psycopg.

Something *unbelievably brutal* we could attempt on our side could be to sneak an ssl structure in the connection after the creation of the structure but before attempting connection to bypass [this branch](https://github.com/postgres/postgres/blob/da505eafca81bc81c34f729521a42a093408cb92/src/interfaces/libpq/fe-secure-openssl.c#L132) but it requires access to the PGconn structure internals, which are outside the public interface.

ISTM that it's possible to add a new function to the libpq, which may be called [between PQconnetStart and PQconnectPoll](https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-PQCONNECTSTARTPARAMS) to initialise the ssl context from the outside (e.g. `bool PQsetSSLContext(PGconn *conn, SSL_CTX *ctx)`).

I would be open to work on the feature and to propose it upstream to the PostgreSQL developers, if the work has some financial support.

**Comment by dvarrazzo** [comment]
Closing the conversation as the ticket #1421 is probably more appropriate to work on.
